### ISSUE TITLE
Interoperability and arrays discussion

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#10632]
Brad pointed out that there hadn't been an issue discussing the addition of the chpl_external_array wrapper as a way to handle passing arrays between Python/C and Chapel in exported function.  The strategy is different than how we handle such types in extern function.  This issue is an attempt to capture that logic, and also to leave an opening for further discussion.

Concerns:
- Chapel's notion of indices is different than C.  Using a c pointer to pass in the array means that the
Chapel function either has to "just know" the length of the array, or also take in a size argument for the array.
  - We wanted the writer of the Chapel exported function to be able to declare arguments as ordinary Chapel arrays, which put the burden of creating the size argument on the compiler generating our wrapper function (and eliminated the ability to "just know" the size - when it is specified in the array argument, you don't have a way to double check that the argument passed had the required length from the array itself, so far as I know).  There were a number of concerns when doing this.
    - Where would the size argument be placed in the list of arguments?
      - This would mean that an exported function with an array argument can't use certain names for its other arguments
      - Additionally, the compiler currently only supports adding arguments to the end or beginning of a call or argument list.  We would need to add support for insertion at a particular place if we wanted to have the size argument follow the array it refers to (for instance).
    - What is the behavior when the function takes in multiple arrays?  What should the size argument be named to avoid conflicting arguments?  If the additional arguments are all placed at the beginning or end of the call/argument list, what order should the user assume?
    - Prior to the generation of the header file for the user, this would have been a gaping source of bugs for the user, as they would have to guess based on our guidance the order of the arguments and their names and types.  With the generated header file, they can look there for how the call should be structured, but may still have to do some guesswork based on what they see.
- Memory management:
  - Arrays created in Chapel need to be free'd using the connected strategy.
    - Again, having the user "just know" what to do with a returned array leaves a lot more room for potential errors
  - Arrays created in C or Python ought to be able to indicate to Chapel how to clean up after themselves, or if Chapel is responsible for doing so.
    - Our current implementation of c_fn_ptr is not really set up well for this, so taking it as an argument would be difficult, if not impossible.

Current implementation:
We chose to make a wrapper type that stores the size and free function, as well as providing a couple of functions that can always be called to create this wrapper given either the number of elements and the type or an already created array and its known size.  We also provide a function that will call the free function stored (if one is stored, and will continue without issue if one is not).  This eliminates the issue with additional arguments and their naming, and allows control over the clean up behavior.  Unfortunately there is no recourse for determining the element type of the array, so some assumptions have to be made there, but overall we feel that this is a better situation.

**Comment by lydia-duncan** [comment]
@mppf - does this seem like an accurate summary of our thought process?

**Comment by mppf** [comment]
yes it seems pretty good to me. I'll add that the other thing I was thinking about was that in the future we might support more exotic NumPy arrays (2D, strided, ...). Having the chpl_external_array  struct will allow us to add fields to the struct to handle the more exotic cases - and that can be done without breaking existing code using 1D arrays (although that code might not work for the 2D arrays...). In contrast, if we found we needed to add an argument (say rank/dimension) to all calls, it would break all the code.

**Comment by bradcray** [comment]
Issue #10636 got to the top of my stack before this one did, but I captured my thoughts / concerns about this issue [in a comment there](https://github.com/chapel-lang/chapel/issues/10636#issuecomment-410786978).
